query-algebrizer/src/clauses/or.rs

```diff
@@ -343,41 +354,381 @@ impl ConjoiningClauses {
+        // variable. We thus generate a set of possible types, and we intersect it with the
+        // types already possible in the CC. If the resultant set is empty, the pattern cannot
+        // match. If the final set isn't unit, we must project a type tag column.
+        self.intersect(folded)
+    }
+
+    fn intersect(&mut self, mut cc: ConjoiningClauses) -> Result<()> {
+        if cc.is_known_empty {
+            self.is_known_empty = true;
+            self.empty_because = cc.empty_because;
+        }
+        self.wheres.append(&mut cc.wheres);
+        self.narrow_types(cc.known_types);
         Ok(())
     }
 }
+
+#[cfg(test)]
+mod testing {
+    extern crate mentat_query_parser;
+
+    use super::*;
+
+    use std::collections::BTreeMap;
+
+    use mentat_core::attribute::Unique;
+    use mentat_core::{
+        Attribute,
+    };
+
+    use mentat_query::{
+        NamespacedKeyword,
+        NonIntegerConstant,
+        Variable,
+    };
+
+    use self::mentat_query_parser::{
+        parse_find_string,
+    };
+
+    use clauses::{
+        add_attribute,
+        associate_ident,
+        ident,
+        unit_type_set,
+    };
+
+    use types::{
+        ColumnConstraint,
+        DatomsTable,
+        QualifiedAlias,
+        QueryValue,
+        SourceAlias,
+    };
+
+    use algebrize;
+
+    fn alg(schema: &Schema, input: &str) -> ConjoiningClauses {
+        let parsed = parse_find_string(input).expect("parse failed");
+        println!("Parsed is {:?}", parsed);
+        algebrize(schema.into(), parsed).expect("algebrize failed").cc
+    }
+
+    fn prepopulated_schema() -> Schema {
+        let mut schema = Schema::default();
+        associate_ident(&mut schema, NamespacedKeyword::new("foo", "name"), 65);
+        associate_ident(&mut schema, NamespacedKeyword::new("foo", "knows"), 66);
+        associate_ident(&mut schema, NamespacedKeyword::new("foo", "parent"), 67);
+        associate_ident(&mut schema, NamespacedKeyword::new("foo", "age"), 68);
+        add_attribute(&mut schema, 65, Attribute {
+            value_type: ValueType::String,
+            multival: false,
+            ..Default::default()
+        });
+        add_attribute(&mut schema, 66, Attribute {
+            value_type: ValueType::String,
+            multival: true,
+            ..Default::default()
+        });
+        add_attribute(&mut schema, 67, Attribute {
+            value_type: ValueType::String,
+            multival: true,
+            ..Default::default()
+        });
+        add_attribute(&mut schema, 68, Attribute {
+            value_type: ValueType::Long,
+            multival: false,
+            ..Default::default()
+        });
+        schema
+    }
+
+    /// Test that if all the attributes in an `or` fail to resolve, the entire thing fails.
+    #[test]
+    fn test_schema_based_failure() {
+        let schema = Schema::default();
```

Comment:
This really threw me.  Can you use attributes that aren't in the prepopulated schema?